data_steward/cdr_cleaner/clean_cdr_engine.py

```diff
@@ -37,129 +41,225 @@ def add_console_logging(add_handler):
-
-        # check if the job is complete and an error has occurred
-        is_errored, error_message = bq_utils.job_status_errored(query_job_id)
-
-        if is_errored:
-            LOGGER.error(
-                format_failure_message(project_id=project,
-                                       statement=statement,
-                                       exception=error_message))
-            failures += 1
-        else:
-            if destination_table is not None:
-                updated_rows = results.get("totalRows")
-                if updated_rows is not None:
-                    LOGGER.info(
-                        f"Query returned {updated_rows} rows for {destination_dataset}.{destination_table}"
-                    )
-
-            successes += 1
-
-    if successes > 0:
-        LOGGER.info(
-            f"Successfully applied {successes} clean rules for {project}.{data_stage}"
-        )
-    else:
-        LOGGER.warning(
-            f"No clean rules successfully applied to {project}.{data_stage}")
+    :param clazz: Clean class or clean function to check the signature of
+    :param kwargs: optional arguments provided by the user
+    :return: filtered dictionary of kwargs
+    :raises ValueError: if a required param for 'clazz' is missing from kwargs
+    """
+    params = inspect.signature(clazz).parameters
+    mandatory_params = ['project_id', 'dataset_id']
+    rule_params = [param for param in params if param not in mandatory_params]
+    # filter kwargs based on required params
+    kwargs = {k: v for k, v in kwargs.items() if k in rule_params}
+    if any(param not in kwargs.keys() for param in rule_params):
+        raise ValueError(
+            f'Params {[param for param in rule_params if param not in kwargs.keys()]} '
+            f'not provided for cleaning rule {clazz.__name__}')
+    return kwargs
 
-    if failures > 0:
-        LOGGER.warning(
-            f"Failed to apply {failures} clean rules for {project}.{data_stage}"
-        )
+
+def infer_rule(clazz, project_id, dataset_id, **kwargs):
+    """
+    Extract information about the cleaning rule
+
+    :param clazz: Clean rule class or old style clean function
+    :param project_id: identifies the project
+    :param dataset_id: identifies the dataset to clean
+    :param kwargs: keyword arguments a cleaning rule may require
+    :return:
+        query_function: function that generates query_list
+        setup_function: function that sets up the tables for the rule
+        rule_info: dictionary of information about the rule
+            keys:
+                query_function: function that generates query_list
+                setup_function: function that sets up the tables for the rule
+                function_name: name of the query function
+                module_name: name of the module containing the function
+                line_no: points to the source line where query_function is
+    """
+    kwargs = validate_params(clazz, **kwargs)
```

Comment:
Is it possible for this call to move up a level? As it stands a missing parameter may only be caught **after** the engine has already visited (e.g. executed) one or more rules. Alternatively, `clean_dataset` and `get_query_list` could collect the full set of parameters required by all rules and proceed **iff** the minimum set are provided. This would save a whole lot of pain.

Reply:
I think it would be appropriate for a function which collects all the params required by a run and compares them with the supplied args to have the word "validate" in its name.

Reply:
I've moved it up a level to `clean_cdr.py` as `validate_all_params`